### ISSUE TITLE
Flatland Replays

### ISSUE THREAD

**Issue by emilioziniades** [emilioziniades/flatland#95]
As per our discussion, we have all the data needed to show the history of flatland visually. Here's a suggested implementation:

- We have a third tab called 'Replay'
- On that tab, there is a Canvas
- Below the canvas is a slider. Far left = beginning of flatland, far right = present
- The user can slide left and right to see how the canvas changes over time
- There is a play button below that, which will replay the whole flatland history 

This is a fun way to visually see the history of the Canvas. Let me know your thoughts

**Comment by gshorer97** [comment]
This will be a great addition! We need to consider how to define the animation increments/steps as well as the time interval between these steps.

Here are some potential options:
(a) Each transaction is a step, and the steps are uniform distance apart.
(b) Each transaction is a step but the steps are spaced apart proportionally to the datetime they occurred (non-uniform intervals)
(c) Each day/week is a step with each step showing the changes from that day/week (uniform intervals but not every transaction is observed) - this option may also be the least resource intensive

**Comment by gshorer97** [comment]
(d) Every n transactions is a step with uniform intervals

**Comment by gshorer97** [comment]
Some added thoughts: 
- I think the canvas should be quite bare bones - just the canvas without the side panel or ability to click on a square.
- Maybe we use styled divs instead of buttons
- on square hover could bring up a tooltip of owner and colour?
- Perhaps it would be nice to have a small event summary feature where there is text appearing and disappearing "narrating" the changes on the canvas (one liner summary) - a very minimal event toast if you will. almost like how an active twitch chat scrolls up to make room for newer comments

**Comment by emilioziniades** [comment]
I think I am leaning towards c) of the options above. Uniform time intervals seem the most intuitive. It would retain the 'replay' nature of the slider, so that simultaneous changes would reflect accurately, instead of being staggered as in a), for example.

Instead of date-time though, I think that block height is a natural choice for the slider scale. They are simple integers which will make them easy to deal with in our code. The minimum value is the earliest block with Flatland events (9996259) and the maximum is the most recent block with Flatland events (10172101 currently).

Let me know what you think of the above.

In terms of the additional comments:

- Totally agree on the barebones approach. Really simple just a canvas of divs with a slider below. 
- The 'twitch chat' toasts is brilliant dude. We could reuse the event toast with a `oneline` prop
- Also yes to transition animation. It might be tricky to implement. I think I'll put together Replay without this to start, and try implement it afterwards
- Not sure the on-hover tooltip is appropriate here. Am in favour of keeping the Replay canvas for that one specific purpose. Information like square ownership is probably more appropriately displayed on the actual Canvas.

**Comment by gshorer97** [comment]
Regarding practicalities:
- Using blockheight is a great idea. Simple and already available. I don't think it's too important to print this information to the user but if we do we should anchor the min value to 0 (i.e. `min value - min value` to `max value - min value`)
- I think the increments still need to be defined - if we are incrementing by the changes for each day then date conversions will still need to occur. We could just increment by `n` transactions each time?
- Agreed to delaying animations. We can scale up as we go: (1) Canvas only, (2) narrations, (3) animations and polish

Regarding philosophy:
- I am happy with limiting all functional/interactive content from the Replay canvas

**Comment by emilioziniades** [comment]
>     * I think the increments still need to be defined - if we are incrementing by the changes for each day then date conversions will still need to occur. We could just increment by `n` transactions each time?

Not sure if I'm understanding you correctly, but I think that by choosing min and max blockheights, the increment is going to automatically be in terms of blockheight. The user can increase one block at a time using the arrow keys (once focused on slider), or just slide it, and the blockheight will update pretty fluidly.

**Comment by emilioziniades** [comment]
Have made some progress in commit df55f849cfe75d8971dc55611cf80a872979603a , so will give an update on progress here:

- [ ] Canvas replay with slider
- [ ] Button for replay playback
- [ ] One line event toast feed 
- [ ] Transition animations & polish

**Comment by gshorer97** [comment]
> Not sure if I'm understanding you correctly, but I think that by choosing min and max blockheights, the increment is going to automatically be in terms of blockheight. The user can increase one block at a time using the arrow keys (once focused on slider), or just slide it, and the blockheight will update pretty fluidly.

You're right. I should've been more clear here and asked what increment we'll use to jump between blockheights. I was initially worried about performance and smoothness of the animation without compressing the changes too drastically. I see now that incrementing the blockheight by 1 is quite fluid (and without data loss).

Thanks for all the work so far! It's looking great

**Comment by emilioziniades** [comment]
Ok great, glad we're on the same page.

I am having issues with the replay button. Currently, the slider's value is controlled by the `range` state variable. `handleChange` updates that `range` variable when a user moves the slider.

However, when I try to change the value of `range` from another part of the component (i.e. the replay button), the `range`   value does change, and the slider updates, however, the Flatland snapshot on the canvas doesn't change. So this means that `handleChange` is not firing when the value of `range` is updated from outside of the Form itself.

**Comment by gshorer97** [comment]
hmm... I havent taken a look at the code in much depth yet. Perhaps you can use a useEffect hook which listens for the range value change? 

Otherwise [this article](https://hustle.bizongo.in/simulate-react-on-change-on-controlled-components-baa336920e04) may have the solution. It discusses how to work with changing states programmatically (which I'm assuming is what is being done with the replay button)

**Comment by emilioziniades** [comment]
Thanks dude this article looks very promising. It's the exact same problem I am having! Will let you know how it goes.

**Comment by emilioziniades** [comment]
Replay button working in commit 04cf755d88ef05a11c590a98cdbed2d41b659ae1 . It's very rough, but it works!
